File: modules/platforms/dotnet/Apache.Ignite.Core.Tests/Cache/Query/Continuous/ContinuousQueryTest.cs
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */


namespace Apache.Ignite.Core.Tests.Cache.Query.Continuous
{
    using System;
    using System.Collections.Concurrent;
    using System.Collections.Generic;
    using Apache.Ignite.Core.Binary;
    using Apache.Ignite.Core.Cache;
    using Apache.Ignite.Core.Cache.Event;
    using Apache.Ignite.Core.Cache.Query.Continuous;
    using Apache.Ignite.Core.Configuration;
    using Apache.Ignite.Core.Log;
    using Apache.Ignite.Core.Tests.Client.Cache;
    using NUnit.Framework;

    /// <summary>
    /// Tests continuous queries with security enabled and disabled.
    /// </summary>
    [TestFixture(false)]
    [TestFixture(true)]
    [Category(TestUtils.CategoryIntensive)]
    public class ContinuousQueryTest
    {
        /** Flag to enable Ignite security. */
        private readonly bool _enableSecurity;

        /** Logger for tracking errors. */
        private readonly ListLogger _logger = new ListLogger(new ConsoleLogger())
        {
            EnabledLevels = new[] {LogLevel.Error}
        };

        /// <summary>
        /// Initializes a new instance of <see cref="ContinuousQueryTest"/>. 
        /// </summary>
        public ContinuousQueryTest(bool enableSecurity)
        {
            _enableSecurity = enableSecurity;
        }
        
        /// <summary>
        /// Clears persistence directory before and after the test.
        /// </summary>
        [TestFixtureSetUp]
        [TestFixtureTearDown]
        public void FixtureTearDown()
        {
            TestUtils.ClearWorkDir();
        }

        /// <summary>
        /// Tests same query on multiple nodes.
        /// This tests verifies that there are no exception on Java side during event delivery.
        /// </summary>
        [Test]
        [Category(TestUtils.CategoryIntensive)]
        public void TestSameQueryMultipleNodes()
        {
            using (var ignite = StartIgnite())
            {
                ignite.GetCluster().SetActive(true);

                var cache = ignite.GetOrCreateCache<Guid, Data>("data");
                cache.QueryContinuous(new ContinuousQuery<Guid, Data>(new Listener()));

                using (var ignite2 = StartIgnite())
                {
                    var cache2 = ignite2.GetOrCreateCache<Guid, Data>("data");
                    cache2.QueryContinuous(new ContinuousQuery<Guid, Data>(new Listener()));

                    for (var i = 0; i < 100; i++)
                    {
                        PutEntry(cache2);
                        PutEntry(cache);
                    }
                }
            }

            Assert.AreEqual(0,
                _logger.Entries.FindAll(e => e.Message.Contains("CacheEntryEventFilter failed")).Count);
        }

        /// <summary>
        /// Puts the entry and verifies events.
        /// </summary>
        private static void PutEntry(ICache<Guid, Data> cache)
        {
            // Put new entry.
            var entry = new Data {Id = Guid.NewGuid()};
            cache.Put(entry.Id, entry);

            // Wait for events.
            TestUtils.WaitForTrueCondition(() => Listener.Events.Count == 2);

            ICacheEntryEvent<Guid, Data> e;

            // Two listeners  - two events.
            Assert.IsTrue(Listener.Events.TryPop(out e));
            Assert.AreEqual(entry.Id, e.Key);

            Assert.IsTrue(Listener.Events.TryPop(out e));
            Assert.AreEqual(entry.Id, e.Key);
        }

        /// <summary>
        /// Starts the ignite.
        /// </summary>
        private IIgnite StartIgnite()
        {
            var ignite = Ignition.Start(new IgniteConfiguration(TestUtils.GetTestConfiguration())
            {
                BinaryConfiguration = new BinaryConfiguration(typeof(Data)),
                AutoGenerateIgniteInstanceName = true,
                DataStorageConfiguration = new DataStorageConfiguration
                {
                    DefaultDataRegionConfiguration = new DataRegionConfiguration
                    {
                        PersistenceEnabled = _enableSecurity,
                        Name = DataStorageConfiguration.DefaultDataRegionName,
                    }
                },
                AuthenticationEnabled = _enableSecurity,
                Logger = _logger,
                IsActiveOnStart = false
            });
            
            ignite.GetCluster().SetBaselineAutoAdjustEnabledFlag(true);

            return ignite;
        }

        private class Data
        {
            public Guid Id;
        }

        private class Listener : ICacheEntryEventListener<Guid, Data>
        {
            public static readonly ConcurrentStack<ICacheEntryEvent<Guid, Data>> Events
                = new ConcurrentStack<ICacheEntryEvent<Guid, Data>>();

            public void OnEvent(IEnumerable<ICacheEntryEvent<Guid, Data>> evts)
            {
                foreach (var e in evts)
                {
                    Events.Push(e);
                }
            }
        }
    }
}

File: modules/core/src/test/java/org/apache/ignite/internal/processors/performancestatistics/ContinuousQueryTest.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.ignite.internal.processors.performancestatistics;

import java.util.Arrays;
import java.util.HashSet;
import java.util.Set;
import java.util.UUID;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.atomic.AtomicLong;
import javax.cache.Cache;
import org.apache.ignite.IgniteCache;
import org.apache.ignite.cache.query.ContinuousQueryWithTransformer;
import org.apache.ignite.cache.query.QueryCursor;
import org.apache.ignite.configuration.IgniteConfiguration;
import org.apache.ignite.internal.IgniteEx;
import org.apache.ignite.internal.IgniteInterruptedCheckedException;
import org.apache.ignite.internal.processors.cache.query.continuous.CacheContinuousQueryManager;
import org.apache.ignite.internal.util.typedef.internal.CU;
import org.apache.ignite.internal.util.typedef.internal.U;
import org.junit.Test;

import static org.apache.ignite.internal.processors.performancestatistics.OperationType.CQ_ENTRY_FILTERED;
import static org.apache.ignite.internal.processors.performancestatistics.OperationType.CQ_ENTRY_PROCESSED;
import static org.apache.ignite.internal.processors.performancestatistics.OperationType.CQ_ENTRY_TRANSFORMED;

/**
 * Tests performance statistics.
 */
public class ContinuousQueryTest extends AbstractPerformanceStatisticsTest {
    /** Ignite. */
    private static IgniteEx srv;

    /** Test cache. */
    private static IgniteCache<Object, Object> cache;

    @Override protected IgniteConfiguration getConfiguration(String igniteInstanceName) throws Exception {
        IgniteConfiguration cfg = super.getConfiguration(igniteInstanceName);

        cfg.setCacheConfiguration(defaultCacheConfiguration());

        return cfg;
    }

    /** {@inheritDoc} */
    @Override protected void beforeTestsStarted() throws Exception {
        srv = startGrid(0);

        cache = srv.cache(DEFAULT_CACHE_NAME);
    }

    /** {@inheritDoc} */
    @Override protected void afterTest() throws Exception {
        super.afterTest();

        cache.clear();
    }

    /** @throws Exception If failed. */
    @Test
    public void testInternalQuery() throws Exception {
        int expFilterDur = 1000;

        cache.put(1, 1);

        CacheContinuousQueryManager contQryMgr = srv.context().cache().context().cacheContext(CU.cacheId(DEFAULT_CACHE_NAME))
            .continuousQueries();

        CountDownLatch evtLatch = new CountDownLatch(1);

        startCollectStatistics();

        long totalDur;

        long startTimeNanos = System.nanoTime();

        contQryMgr.executeInternalQuery(
            evts -> evts.forEach(e -> evtLatch.countDown()),
            e -> {
                try {
                    U.sleep(expFilterDur);
                }
                catch (IgniteInterruptedCheckedException ignored) {
                    // No-op.
                }

                return true;
            }, true, true, false, false);

        evtLatch.await();

        totalDur = System.nanoTime() - startTimeNanos;

        Set<OperationType> expOps = new HashSet<>(Arrays.asList(CQ_ENTRY_FILTERED, CQ_ENTRY_PROCESSED));

        AtomicLong filterDur = new AtomicLong();
        AtomicLong processDur = new AtomicLong();

        stopCollectStatisticsAndRead(new TestHandler() {
            @Override public void continuousQueryOperation(UUID nodeId, OperationType type, UUID routineId,
                long opStartTime, long duration, int entCnt) {
                assertTrue(expOps.remove(type));

                if (type == CQ_ENTRY_FILTERED)
                    filterDur.set(duration);

                if (type == CQ_ENTRY_PROCESSED)
                    processDur.set(duration);
            }
        });

        assertTrue(expOps.isEmpty());

        assertTrue(processDur.get() + filterDur.get() <= totalDur);
    }

    /** @throws Exception If failed. */
    @Test
    public void testLocalQueryWithTransformer() throws Exception {
        int expTransformDur = 1000;

        CountDownLatch evtLatch = new CountDownLatch(1);

        ContinuousQueryWithTransformer<Integer, Integer, Integer> qry = new ContinuousQueryWithTransformer<>();

        qry.setLocal(true);

        qry.setLocalListener(evts -> evts.forEach(e -> evtLatch.countDown()));

        qry.setRemoteTransformerFactory(() -> e -> {
            try {
                U.sleep(expTransformDur);
            }
            catch (IgniteInterruptedCheckedException ignored) {
                // No-op.
            }

            return e.getValue();
        });

        startCollectStatistics();

        long totalDur;

        try (QueryCursor<Cache.Entry<Integer, Integer>> cur = cache.query(qry)) {
            long startTimeNanos = System.nanoTime();

            cache.put(1, 1);

            evtLatch.await();

            totalDur = System.nanoTime() - startTimeNanos;
        }

        Set<OperationType> expOps = new HashSet<>(Arrays.asList(CQ_ENTRY_PROCESSED, CQ_ENTRY_TRANSFORMED));

        AtomicLong transformDur = new AtomicLong();
        AtomicLong processDur = new AtomicLong();

        stopCollectStatisticsAndRead(new TestHandler() {
            @Override public void continuousQueryOperation(UUID nodeId, OperationType type, UUID routineId,
                long opStartTime, long duration, int entCnt) {
                assertTrue(expOps.remove(type));

                if (type == CQ_ENTRY_PROCESSED)
                    processDur.set(duration);

                if (type == CQ_ENTRY_TRANSFORMED)
                    transformDur.set(duration);
            }
        });

        assertTrue(expOps.isEmpty());

        assertTrue(processDur.get() + transformDur.get() <= totalDur);
    }
}